### PR TITLE
Improvement of FFI data zeroizing.

### PR DESCRIPTION
**Pull request by Superhepper**
- Adds a new InPlaceFfiDataZeroizer trait that types can implement in order to provide a zeroize capabilities for their associated FFI data.
- Implements the new InPlaceFfiDataZeroizer for several types in order to improve zerozation whe calling the create context method.
- Adds CreateInputHandler type to manage the input parameters the create context method and to zeroize the FFI data when dropped.
- Adds CreateOutputHandler to manage the FFI output parameters and when converted into CreateKeyResult properly zeroize FFI data allocated by the TSS.

### CODE REVIEW

File: tss-esapi/src/ffi_data.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// Copyright 2021 Contributors to the Parsec project.
```
Comment (ionut-arm) [inline tss-esapi/src/ffi_data.rs:1]:
```suggestion
// Copyright 2022 Contributors to the Parsec project.
```

File: tss-esapi/src/ffi_data.rs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+pub(crate) fn to_owned_with_zeroized_source<T, U>(data_ptr: *mut T) -> T
+where
+    T: Copy,
+    U: InPlaceFfiDataZeroizer<T>,
+{
+    let mut ffi_data = unsafe { MBox::from_raw(data_ptr) };
+    let owned_ffi_data: T = *ffi_data;
+    U::zeroize_ffi_data_in_place(ffi_data.deref_mut());
+    owned_ffi_data
+}
```
Comment (ionut-arm) [inline tss-esapi/src/ffi_data.rs:19]: Hmm, this feels like something MBox should be able to do... 😞

File: tss-esapi/src/structures/tagged/symmetric.rs [hunk 3]
```diff
@@ -1,10 +1,13 @@
+use zeroize::Zeroize;
+
```
Comment (ionut-arm) [inline tss-esapi/src/structures/tagged/symmetric.rs:2]: The import is before the copyright o_O

**Review by ionut-arm — COMMENTED** [review]
Thanks, looks quite nice!

One thing, though, shouldn't we implement something like zeroizing-on-drop using this new trait for stuff like [`PrivateKeyRsa`](https://docs.rs/tss-esapi/7.1.0/tss_esapi/structures/struct.PrivateKeyRsa.html)?

**Comment by Superhepper** [comment]
> Thanks, looks quite nice!
> 
> One thing, though, shouldn't we implement something like zeroizing-on-drop using this new trait for stuff like [`PrivateKeyRsa`](https://docs.rs/tss-esapi/7.1.0/tss_esapi/structures/struct.PrivateKeyRsa.html)?

We absolutely should. I just did it for the items in the ```create``` context method to provide an example. If we think this is a way to go then I will add more commits that implements things for the other structs as well.

File: tss-esapi/src/ffi_data.rs (hunk 2 above)
Comment (Superhepper) [reply tss-esapi/src/ffi_data.rs:19]: I agree though the MBox crate does not seem to be maintained any more.

**Comment by Superhepper** [comment]
@ionut-arm, @wiktor-k 

Do you think that this draft shows the way we should take?

**Comment by ionut-arm** [comment]
> Do you think that this draft shows the way we should take?

I think I'm happy with the approach, but I'll want to review it again, just to make sure I get all the details. I'm also mulling some changes related to our use of `MBox`, I'll write that up in an issue.

**Comment by ionut-arm** [comment]
Damn, forgot about reviewing it again - please move it to "ready for review", I'll add it to my agenda and we can get it merged soon.

**Comment by Superhepper** [comment]
I was not really looking for a review just yet more like if you approved of the way things was handled.

**Comment by ionut-arm** [comment]
I've looked at it again - I think one question I have is why do stuff like `impl InPlaceFfiDataZeroizer<TPMT_SYM_DEF_OBJECT> for SymmetricDefinitionObject` instead of just `impl InPlaceFfiDataZeroizer for TPMT_SYM_DEF_OBJECT`? Even though the FFI type isn't owned by this crate, the trait is so it should work. It could then act essentially as a constructor for those FFI types.

**Comment by Superhepper** [comment]
> I've looked at it again - I think one question I have is why do stuff like `impl InPlaceFfiDataZeroizer<TPMT_SYM_DEF_OBJECT> for SymmetricDefinitionObject` instead of just `impl InPlaceFfiDataZeroizer for TPMT_SYM_DEF_OBJECT`? Even though the FFI type isn't owned by this crate, the trait is so it should work. It could then act essentially as a constructor for those FFI types.

I did it this way because I thought it would break the 'Orphan rule' to implement the trait I defined for a type from another crate. But I realise this might not be the case.

**Pull request closed without merging**